polkadot/xcm/pallet-xcm/src/precompiles.rs

```diff
@@ -0,0 +1,530 @@
+use IXcm::IXcmCalls;
+
+const LOG_TARGET: &str = "xcm::precompiles";
+
+// Revert macro to handle errors in a consistent way across the precompile.
+#[macro_export]
+macro_rules! revert {
+    ($target:expr, $error:expr, $($arg:tt)*) => {{
+        let message = format!($($arg)*);
+        error!(target: $target, error = ?$error, "{}", message);
+        Error::Revert(message.into())
+    }};
+}
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+impl<Runtime> Precompile for XcmPrecompile<Runtime>
+where
+	Runtime: crate::Config + pallet_revive::Config,
+{
+	type T = Runtime;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) =>
+				frame_system::RawOrigin::Signed(account_id.clone()).into(),
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let _ = env.charge(<Runtime as Config>::WeightInfo::send())?;
+
+				let final_destination = VersionedLocation::decode_all(&mut &destination[..])
+					.map_err(|error| {
+						revert!(LOG_TARGET, error, "XCM send failed: Invalid destination format")
+					})?;
+
+				let final_message = VersionedXcm::<()>::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					revert!(LOG_TARGET, error, "XCM send failed: Invalid message format")
+				})?;
+
+				crate::Pallet::<Runtime>::send(
+					frame_origin,
+					final_destination.into(),
+					final_message.into(),
+				)
+				.map(|message_id| message_id.encode())
+				.map_err(|error| {
+					revert!(
+						LOG_TARGET,
+						error,
+						"XCM send failed: destination or message format may be incompatible"
+					)
+				})
+			},
+			IXcmCalls::xcmExecute(IXcm::xcmExecuteCall { message, weight }) => {
+				let weight = Weight::from_parts(weight.refTime, weight.proofSize)
+					.saturating_add(<Runtime as Config>::WeightInfo::execute());
+				let charged_amount = env.charge(weight)?;
+
+				let final_message = VersionedXcm::decode_all_with_depth_limit(
+					MAX_XCM_DECODE_DEPTH,
+					&mut &message[..],
+				)
+				.map_err(|error| {
+					revert!(LOG_TARGET, error, "XCM execute failed: Invalid message format")
+				})?;
+
+				let result =
+					crate::Pallet::<Runtime>::execute(frame_origin, final_message.into(), weight);
+
+				let pre = DispatchInfo {
+					call_weight: weight,
+					extension_weight: Weight::zero(),
+					..Default::default()
+				};
+
+				// Adjust gas using actual weight or fallback to initially charged weight
+				let actual_weight = frame_support::dispatch::extract_actual_weight(&result, &pre);
```

Comment:
```suggestion
				let actual_weight = frame_support::dispatch::extract_actual_weight(&result, &pre).saturating_add(<Runtime as Config>::WeightInfo::execute());
```

Reply:
This is already added at the start of the function

Reply:
For the pre-dispatch weight. But it also has to be added when calculating the actual weight. You need to use the same formula for both weights.